### PR TITLE
Provide more explanation for Deref in String docs

### PR DESCRIPTION
**Pull request by natboehm**
While working on a different project I encountered a point of confusion where using `&String` to dereference a `String` into `&str` did not compile. I found the explanation of [String Deref](https://doc.rust-lang.org/std/string/struct.String.html#deref), thought that it matched what I was trying to do, and was confused as to why my program did not compile when the docs stated that it would work with 'any function which takes a `&str`'. At the bottom it is mentioned that this will 'generally' work, unless `String` is needed, but I found this statement confusing based on the previous claim of 'any'. Looking further into the docs I was able to find the function `as_str()` that works instead. 

I thought it might be helpful to mention here deref coercion, an instance in which using `&String` does not work, to explain why it does not work, then direct users to a different option that should work in this instance. A user casually skimming the page will likely come to this explanation first, then find `as_str()` later, but be no the wiser as to what potentially went wrong.

r? @steveklabnik

### CODE REVIEW

File: src/liballoc/string.rs [hunk 1]
```diff
@@ -160,8 +160,33 @@ use boxed::Box;
...
+///
+/// fn main() {
+///     let example_string = String::from("example_string");
+///     example_func(&example_string);
+/// }
+/// ```
+///
+/// What would work in this case is changing the line
+/// `example_func(&example_string);` to 
+/// `example_func(example_string.to_str());`. This works because we're doing
```
Comment (kennytm) [inline src/liballoc/string.rs]: I think the idiomatic solution is call it like this:

```rust
    example_func(&*example_string);
```

You *dereference* a `String` into a `str`, and then reference the `str` back to a `&str`.

File: src/liballoc/string.rs [hunk 2]
```diff
@@ -160,8 +160,33 @@ use boxed::Box;
...
+/// fn example_func<A: TraitExample>(example_arg: A) {}
+///
+/// fn main() {
+///     let example_string = String::from("example_string");
+///     example_func(&example_string);
+/// }
+/// ```
+///
+/// What would work in this case is changing the line
+/// `example_func(&example_string);` to
```
Comment (kennytm) [inline src/liballoc/string.rs]: This line has a trailing whitespace making CI to reject the commit.

```
[00:03:44] tidy error: /checkout/src/liballoc/string.rs:187: trailing whitespace
[00:03:45] some tidy checks failed
```

File: src/liballoc/string.rs (hunk 1 above)
Comment (steveklabnik) [inline src/liballoc/string.rs]: I do think that `&*` is slightly more idiomatic, but it's also harder to understand. I think showing both is probably a good idea.

Comment (natboehm) [inline src/liballoc/string.rs]: Cool, I'll make the change. I like the idea of having both options in there, thanks!

File: src/liballoc/string.rs [hunk 3]
```diff
@@ -160,8 +160,38 @@ use boxed::Box;
 ///
 /// This will create a [`&str`] from the `String` and pass it in. This
 /// conversion is very inexpensive, and so generally, functions will accept
-/// [`&str`]s as arguments unless they need a `String` for some specific reason.
+/// [`&str`]s as arguments unless they need a `String` for some specific
+/// reason.
 ///
+/// In certain cases Rust doesn't have enough information to make this
+/// conversion, known as deref coercion. In the following example a string
```
Comment (steveklabnik) [inline src/liballoc/string.rs]: this should be

    `Deref`

since it's the name of a trait

**Review by steveklabnik — COMMENTED** [review]
Sorry, one last tiny nit! r=me after this is fixed

**Pull request merged**